src/main/java/org/springframework/data/redis/cache/DefaultRedisCacheWriter.java

```diff
@@ -149,6 +137,103 @@ public byte[] get(String name, byte[] key, @Nullable Duration ttl) {
 		return result;
 	}
 
+	@Override
+	public Mono<byte[]> retrieve(String name, byte[] key, @Nullable Duration ttl) {
+
+		Assert.notNull(name, "Name must not be null");
+		Assert.notNull(key, "Key must not be null");
+
+		Mono<byte[]> result = nonBlockingExecutionStrategy(name).apply(key, ttl);
+
+		result = result.doOnSuccess(byteBuffer -> {
+			if (byteBuffer != null) {
+				statistics.incHits(name);
+			}
+			else {
+				statistics.incMisses(name);
+			}
+		}).doFirst(() -> statistics.incGets(name));
+
+		return result;
+	}
+
+	private BiFunction<byte[], Duration, Mono<byte[]>> nonBlockingExecutionStrategy(String cacheName) {
+		return isReactiveAvailable() ? reactiveExecutionStrategy(cacheName) : asyncExecutionStrategy(cacheName);
+	}
+
+	// Execution Strategy applied when Jedis (non-Reactive driver) is used.
+	// Treats API consistently (that is, using Reactive types, such as Mono) at the RedisCacheWriter level
+	// whether "technically Reactive" or not; clearly Jedis is not "Reactive".
+	private BiFunction<byte[], Duration, Mono<byte[]>> asyncExecutionStrategy(String cacheName) {
```

Comment:
2 things. 

First, my initial reaction was how to support our **Jedis** users, too. So this was the design I came up with.

Second, Spring Framework's contract for the `Cache.retrieve(..)` operation (e.g. [Javadoc](https://docs.spring.io/spring-framework/docs/6.1.0-SNAPSHOT/javadoc-api/org/springframework/cache/Cache.html#retrieve(java.lang.Object))) is very clear... it must not block.

In fact, the approach I took (using `CompletableFuture.supplyAsync(..)` for Jedis) is precisely how Spring Framework handles asynchronous, non-blocking behavior in `Cache.retrieve(..)` for caching providers that do not support asynchronous, non-blocking operations, such as `ConcurrentMapCache` ([source](https://github.com/spring-projects/spring-framework/blob/v6.1.0-M5/spring-context/src/main/java/org/springframework/cache/concurrent/ConcurrentMapCache.java#L159-L171)). In contrast, if the underlying caching provider, such as Caffeine, supports asynchronous, non-blocking operations (in other words, return types, such as `CompletableFuture`), then the implementation is trivial ([source](https://github.com/spring-projects/spring-framework/blob/v6.1.0-M5/spring-context-support/src/main/java/org/springframework/cache/caffeine/CaffeineCache.java#L138-L152)).

Reply:
So, in conclusion, I would say.

Either, we support Jedis users by wrapping the simple cache op in an asynchronous manner, not unlike `ConcurrentMapCache`, or we do not support Jedis at all.

It would be wrong to support Jedis, but only in a blocking manner, which would violate the contract of `Cache.retrieve(..)`.

I am leaning towards supporting Jedis users where possible and the implementation is rather simple.

Reply:
> Either, we support Jedis users by wrapping the simple cache op in an asynchronous manner, not unlike ConcurrentMapCache, or we do not support Jedis at all.

We decided against pretending our blocking infrastructure would support asynchronicity to avoid raising false expectations. `RestTemplate`, JPA, `JdbcTemplate` are several examples that have followed this pattern.

It is not wise to do things just because we can.

> It would be wrong to support Jedis, but only in a blocking manner, which would violate the contract of Cache.retrieve(..).

`retrieve` is allowed to throw `UnsupportedOperationException`. The JCache Cache doesn't support async operations.

Reply:
> _It is not wise to do things just because we can._

No need to state the obvious. Also, I was not part of the original/initial "decision" so I am largely unaware and unfamiliar with what the intended direction is for Spring Data Redis. I am simply acting in the best interest of our community, and I am simply stating my case in favor of this approach, not because we "can".

Also, many caching providers don't support async caching operations yet. Neither did the `ConcurrentMapCache` implementation in the core Spring Framework, but support was added for async caching behavior none-the-less.

The _JCache_ library and overall API is an SPI (a specification, hence JSR-107), not a implementation or caching provider. This is not unlike Spring Framework's Cache Abstraction itself. As such, it is necessarily the lowest common denominator across an ecosystem of caching providers. By way of example, _Caffeine_, which provides async caching support, is also a JCache caching [provider implementation](https://github.com/ben-manes/caffeine/wiki/JCache).

Anyway, I happy to go in whatever direction we decide together.